### PR TITLE
Virts 1014b object schemas & API docs (2)

### PR DESCRIPTION
**Pull request by unkempthenry**
Changes: 

* introduces marshmallow to provide schemas for objects ( https://marshmallow.readthedocs.io/en/stable/ )
* This PR only adds a schema for Agent (more can be added later). The schema is used for serialization in the Agent.display() method, and deserialization in the new Agent.from_dict method.  
* Input validation for agent updates returns 400 Bad Request when input isn't loadable by the schema. 
* Some tests for querying agent and updating agent information via the rest_core endpoint.

Thoughts: 

In addition to simplifying the serialization / deserialization of objects and adding validation to API inputs, having defined schemas for objects will be helpful to generate API documentation that stays up to date.

### CODE REVIEW

File: app/service/rest_svc.py [hunk 1]
```diff
@@ -134,9 +134,10 @@ def __init__(self):
         return dict(contacts=self.get_service('contact_svc').report.get(contact.get('contact'), dict()))
 
     async def update_agent_data(self, data):
-        if 'paw' not in data:
+        paw = data.pop('paw', None)
```
Comment (wbooth) [inline app/service/rest_svc.py]: move this small change to a PR

File: app/objects/c_agent.py [hunk 2]
```diff
@@ -55,6 +73,11 @@ def __init__(self, sleep_min, sleep_max, watchdog, platform='unknown', server='u
         self.contact = contact
         self.access = self.Access.BLUE if group == 'blue' else self.Access.RED
 
+    @classmethod
+    def from_dict(cls, dict_obj):
+        """ Creates an Agent object from parameters stored in a dict. AgentSchema is used to validate inputs."""
+        return cls(**cls.AgentSchema().load(dict_obj))
```
Comment (wbooth) [inline app/objects/c_agent.py]: I'd drop the tests from the PR and propose the marshmallow change first

File: app/objects/c_agent.py [hunk 3]
```diff
@@ -97,11 +120,9 @@ def store(self, ram):
...
 
     async def gui_modification(self, **kwargs):
-        self.update('group', kwargs.get('group'))
-        self.update('trusted', kwargs.get('trusted'))
-        self.update('sleep_min', int(kwargs.get('sleep_min')))
-        self.update('sleep_max', int(kwargs.get('sleep_max')))
-        self.update('watchdog', int(kwargs.get('watchdog')))
+        loaded = self.AgentSchema(only=('group', 'trusted', 'sleep_min', 'sleep_max', 'watchdog')).load(kwargs)
+        for k, v in loaded.items():
+            self.update(k, v)
```
Comment (wbooth) [inline app/objects/c_agent.py]: yea this makes things easier

Comment (unkempthenry) [reply app/objects/c_agent.py]: so no further change to this line, then?

File: app/service/rest_svc.py (hunk 1 above)
Comment (unkempthenry) [reply app/service/rest_svc.py]: https://github.com/mitre/caldera/pull/1427

**Pull request merged**